### PR TITLE
[Backport 1.18] Enable Windows workers (#17555)

### PR DESCRIPTION
**Pull request by davinci26**
Commit Message:

Backport https://github.com/envoyproxy/envoy/commit/1260c5c2d1c5b85639a8e566a6b153198ddbf109 to 1.18

Fixing an issue where Envoy workers are not picking up connections on Windows.

The root cause of the issue is in the Windows kernel.

There are two issues that we need to consider:

If we want to listen from a duplicated socket then we need to duplicate it after we call listen on the original socket.
If duplicated sockets try to accept at the same time, then one of the accept calls might block. Even if the sockets are non-blocking.
The best way to work around that issue is to only listen/accept connections from the first worker thread and then use ExactConnectionBalance to dispatch the connection to another worker thread. This is a temporary solution until the underlying issue is fixed on Windows.

Signed-off-by: Sotiris Nanopoulos <sonanopo@microsoft.com>

Risk Level: Low
Testing: Manual testing and tests
Docs Changes: Added
Release Notes: Added a bug fix note
Platform Specific Features: N/A

### CODE REVIEW

**Comment by davinci26** [comment]
cc @wrowe

File: source/server/listener_impl.cc [hunk 1]
```diff
@@ -523,6 +523,19 @@ void ListenerImpl::buildFilterChains() {
...
   // TCP specific setup.
   if (connection_balancer_ == nullptr) {
+#ifdef WIN32
+    // On Windows we use the exact connection balancer to dispatch connections
+    // from worker 1 to all workers. This is a perf hit but it is the only way
+    // to make all the workers do work.
+    // TODO(davinci26): We can be faster here if we create a balancer implementation
+    // that dispatches the connection to a random thread.
+    ENVOY_LOG(warn,
+              "ExactBalance was forced enabled for TCP listener '{}' because "
```
Comment (dmitri-d) [inline source/server/listener_impl.cc:533]: nit: force enabled

**Comment by dmitri-d** [comment]
This lgtm, but with a fat warning that I don't know much about Windows idiosyncrasies.

**Comment by wrowe** [comment]
@mathetake was correct, need to merge to branch one more time here @davinci26 and I think it will pass CI to be merged.

**Review by wrowe — APPROVED** [review]
LGTM, it would be nice to have several working examples to compare between 1.18/1.19/main evolution of the rest of the  code. TYVM for the backport submission.

**Pull request merged**